File: jbake-core/src/main/java/org/jbake/template/DelegatingTemplateEngine.java
package org.jbake.template;

import org.apache.commons.configuration2.CompositeConfiguration;
import org.jbake.app.ContentStore;
import org.jbake.app.FileUtil;
import org.jbake.app.configuration.JBakeConfiguration;
import org.jbake.template.model.TemplateModel;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.Writer;

import static org.jbake.app.configuration.PropertyList.PAGINATE_INDEX;

/**
 * A template which is responsible for delegating to a supported template engine,
 * based on the file extension.
 *
 * @author Cédric Champeau
 */
public class DelegatingTemplateEngine extends AbstractTemplateEngine {
    private static final Logger LOGGER = LoggerFactory.getLogger(DelegatingTemplateEngine.class);

    private final TemplateEngines renderers;

    /**
     * @deprecated Use {@link #DelegatingTemplateEngine(ContentStore, JBakeConfiguration)} instead.
     *
     * @param config the {@link CompositeConfiguration} of jbake
     * @param db the {@link ContentStore}
     * @param destination the destination path
     * @param templatesPath the templates path
     */
    @Deprecated
    public DelegatingTemplateEngine(final CompositeConfiguration config, final ContentStore db, final File destination, final File templatesPath) {
        super(config, db, destination, templatesPath);
        this.renderers = new TemplateEngines(this.config, db);
    }

    public DelegatingTemplateEngine(final ContentStore db, final JBakeConfiguration config) {
        super(config, db);
        this.renderers = new TemplateEngines(config, db);
    }

    @Override
    public void renderDocument(final TemplateModel model, final String templateName, final Writer writer) throws RenderingException {
        model.setVersion(config.getVersion());
        model.setConfig(config.asHashMap());

        // if default template exists we will use it
        File templateFolder = config.getTemplateFolder();
        File templateFile = new File(templateFolder, templateName);
        String theTemplateName = templateName;
        if (!templateFile.exists()) {
            LOGGER.info("Default template: {} was not found, searching for others...", templateName);
            // if default template does not exist then check if any alternative engine templates exist
            String templateNameWithoutExt = templateName.substring(0, templateName.length() - 4);
            for (String extension : renderers.getRecognizedExtensions()) {
                templateFile = new File(templateFolder, templateNameWithoutExt + "." + extension);
                if (templateFile.exists()) {
                    LOGGER.info("Found alternative template file: {} using this instead", templateFile.getName());
                    theTemplateName = templateFile.getName();
                    break;
                }
            }
        }
        String ext = FileUtil.fileExt(theTemplateName);
        AbstractTemplateEngine engine = renderers.getEngine(ext);
        if (engine != null) {
            engine.renderDocument(model, theTemplateName, writer);
        } else {
            LOGGER.error("Warning - No template engine found for template: {}", theTemplateName);
        }
    }
}

File: jbake-core/src/main/kotlin/org/jbake/template/DelegatingTemplateEngine.kt
package org.jbake.template

import org.jbake.app.ContentStore
import org.jbake.app.FileUtil
import org.jbake.app.RenderingException
import org.jbake.app.configuration.JBakeConfiguration
import org.jbake.template.model.TemplateModel
import org.jbake.util.Logging.logger
import org.slf4j.Logger
import java.io.Writer

/**
 * A template engine which delegates to a supported template engine, based on the file extension.
 */
class DelegatingTemplateEngine(db: ContentStore, config: JBakeConfiguration) : AbstractTemplateEngine(config, db) {
    private val renderers: TemplateEngines = TemplateEngines(config, db)

    @Throws(RenderingException::class)
    override fun renderDocument(model: TemplateModel, templateName: String, writer: Writer) {
        model.version = (config.jbakeVersion)
        model.config = run {
            // Use configuration's asHashMap which provides underscore-style keys and defaults similar to legacy behavior.
            val base: MutableMap<String, Any> = config.asHashMap()

            // Also expose dotted-key aliases for code that prefers dotted keys
            val m: MutableMap<String, Any> = HashMap(base)
            base.forEach { (k, v) -> m.putIfAbsent(k.replace('_', '.'), v) }
            m
        }
        // Also keep a reference to the original configuration object so typed extractors can access the configuration with its original dotted keys.
        // ModelExtractorAdapter will prefer this when reconstructing a RenderContext.
        model["jbake_config"] = config

        // If default template exists we will use it.
        val templateDir = config.templateDir
        var templateFile = templateDir.resolve(templateName)
        var templateFileName = templateName

        if (!templateFile.exists()) {
            log.info("Default template: $templateName was not found, searching for others...")
            // if default template does not exist then check if any alternative engine templates exist
            val templateNameWithoutExt = templateName.dropLast(4)
            for (extension in renderers.recognizedExtensions) {
                templateFile = templateDir.resolve("$templateNameWithoutExt.$extension")
                if (templateFile.exists()) {
                    log.info("Found alternative template file: ${templateFile.getName()} using this instead.")
                    templateFileName = templateFile.getName()
                    break
                }
            }
        }

        val ext = FileUtil.fileExt(templateFileName)
        val engine = renderers.getEngine(ext)
        if (engine == null) {
            log.error("Warning - No template engine found for template: $templateFileName.")
            return
        }

        // Convert Temporals to java.util.Date for all template engines
        val convertedModel = org.jbake.util.convertTemporalsInModelToJavaUtilDate(model)

        engine.renderDocument(convertedModel, templateFileName, writer)
    }

    private val log: Logger by logger()
}
